Add continuous aerosols support

WalterKolczynski-NOAA opened the pull request:
**Description**
Adds support for continuous aerosols when running with GOCART (using the forecast aerosol tracers from the previous cycle as initial tracer fields). This adds a new task to the workflow mesh in forecast-only mode. The new task does not run in the first cycle, though it will appear in rocotostat/rocotoviewer. In subsequent cycles, the task will launch as soon as the needed restart files from the previous cycle are available and gfs_init has completed. There is _currently_ no mechanism to have aerosol_init run for the first cycle (for instance, continuing from a previous experiment); users would have to run the scripts off-line to add aerosols to the initial conditions.

To support the introduction of the necessary python scripts, a new python environment is added to provide necessary libraries. This virtual environment is temporarily being housed in personal space, but an [issue](NOAA-EMC/hpc-stack#408) has been opened with hpc-stack to add it to the standard stack installations. The introduction of miniconda also caused the conversion of all of the rocoto entry scripts (the ones rocoto calls to run the job) to bash. Most of them had been in korne.

To make sure the needed restart files are produced, [config.fcst](https://github.com/WalterKolczynski-NOAA/global-workflow/blob/feature/continuous_aerosols/parm/config/config.fcst#L267) now makes sure the forecast cadence is included in the list of restart times. This may need more testing to ensure it works properly with other restart lists. As part of this, STEP_GFS was updated to properly determine the step from gfs_cyc instead of being a set value.

Closes #366
Fixes #516
Fixes #630

**Type of change**
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**How Has This Been Tested?**
- [x] Forecast-only test on Hera
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

WalterKolczynski-NOAA committed:
Add aerosol continuity

Makes the necessary additions to allow aerosol fields to be retained
from cycle to cycle. Aerosols are taken from the restart file valid at
the initial time from the previous cycle and added to the current
cycle's initial conditions (with some modifications to maintain mass
conservation).

This adds a job to the workflow when aerosols are on. The job will not
run in the first cycle.

In order to make sure there is a restart file written at the valid time,
the forecast cadence is added to the restart_interval list if it is not
already present. This also required fixing STEP_GFS to be appropriately
derived from gfs_cyc. To facilitate this, when the restart_interval is 0,
that is replaced with the max forecast time (which is what 0 is a short-
cut for).

These additions are as-yet untested beyond workflow generation, so there
will undoubtedly be additional commits with fixes.

Refs: #516

WalterKolczynski-NOAA committed:
Change file pattern for IC location in aerosol init

Updated the IC location in aerosol init to the correct location in COM.

Also updated the script to do the modifications in a temporary file
instead of the originals and then replace the originals when complete.
This will avoid issues if the merger fails.

Refs: #516

WalterKolczynski-NOAA committed:
Remove special handling of NetCDF import error in aerosol merge

Removes the special handling of the NetCDF import in the aerosol merge script
and allows the exception to propagate up the chain instead. There really isn't
a need for these import errors to be handled differently than any other.

Ref: #516

WalterKolczynski-NOAA committed:
Close atm tile before using ncks

Closes the NetCDF handle on the atm file after the new variables are added.
The ncks command to remove ntracers so it can be redefined to the new number
of tracers was failing because the file was still open in the python script.

Refs: #516

WalterKolczynski-NOAA committed:
Remove checksum from aerosol-modified tiles

Removes the checksum from the tiile files after the new aerosol fields have
been added, since the checksum is no longer valid and this causes the forecast
to fail.

Refs: #516

WalterKolczynski-NOAA committed:
Fix and refine dependencies for aerosol init

Updates the aerosol init dependencies to use the correct filepath for files
from the previous cycle. Because the needed restart files are held in the
ROTDIR for the previous cycle, but the file name uses the valid time (i.e.
the current cycle), two different cyclestr tags are needed in a single
dependency. The add_dependency function doesn't currently support this, so
the dependency is constructed by hacking the end datadep tag off the first
cyclestr, the leading datadep tag off the second cyclestr, and concatenating
them together to form a single datadep tag.

Also added a dependency to allow the forecast to run without waiting for the
aerosol init job in the first cycle (when the aerosol job doesn't run) and
corrected some other typos/mistakes.

Refs: #516

WalterKolczynski-NOAA committed:
Update all job entry rocoto scripts to use bash

Converted all of the job entry rocoto scripts (the scripts called by rocoto
tasks when the job runs) to bash. Most (all?) of them were in korn and this
was causing issues with loading the new miniconda environment.

Refs: #397, #516

WalterKolczynski-NOAA committed:
Add new miniconda environment on Hera

Miniconda and a new python environment are added to support libraries needed for
python scripts. The stack for these are temporarily built in personal space on
Hera, but an Issue is open with hpc-stack (NOAA-EMC/hpc-stack#408) to add the
new environment to the official stack installations.

This commit *breaks* MET verification on Hera for now. We are working with the
verification team to remedy.

Refs: #516

WalterKolczynski-NOAA commented:
This PR is technically orthogonal to PR #681, but this will be held until that PR is merged and this PR is retested with those changes.

This new python environment has also caused downstream issues that cause the MET verify jobs to fail due to being unable to load its python environment. I am working with @malloryprow to see if we can remove the custom module load done in verify or otherwise mitigate this issue. Until that is fixed, users should switch DO_VRFY to NO in [config.base](https://github.com/WalterKolczynski-NOAA/global-workflow/blob/feature/continuous_aerosols/parm/config/config.base.emc.dyn#L58) until the problem is corrected.

WalterKolczynski-NOAA commented:
Currently I've only built the python environment on Hera, but I'll be expanding to other tier-1 machines soon (unless hpc-stack beats me to it). But for now, this will only work on Hera.

bbakernoaa commented:
@WalterKolczynski-NOAA I got this compiled and running but it failed on the gfsaerosol_init on the second forecast day.  I'm going to keep looking into it but this is the error right now. Specifically, it is failing in the `merge_fv3_aerosol_tile.py` script where the `ncatted` call occurs. 

```bash
Updating ntracer
nco_err_exit(): ERROR Short NCO-generated message (usually name of function that triggered error): nco__open()
nco_err_exit(): ERROR Error code is -101. Translation into English with nc_strerror(-101) is "NetCDF: HDF error"
ERROR: nco__open() unable to open file "/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3test3/gfs.20190802/00/atmos/INPUT/gfs_data.tile1.nc.tmp"
ERROR NC_EHDFERR Error at HDF5 layer
HINT: NC_EHDFERR errors indicate that the HDF5-backend to netCDF is unable to perform the requested task. NCO can receive this devilishly inscrutable error for a variety of possible reasons includ\
ing: 1) The run-time dynamic linker attempts to resolve calls from the netCDF library to the HDF library with an HDF5 libhdf5.a that is incompatible with the version used to build NCO and netCDF. \
2) The file system does not allow the HDF5 flock() function, as of HDF5 1.10.x, to enable multiple processes to open the same file for reading, a feature known as SWMR (Single Write Multiple Read)\
. The fix is to disable the HDF5 flock() by setting an environment variable thusly: "export HDF5_USE_FILE_LOCKING=FALSE". 3) An incorrect netCDF4 library implementation of a procedure (e.g., nc_re\
name_var()) in terms of HDF function calls (e.g., HDF5Lmove()) manifests an error or inconsistent state within the HDF5 layer. This often occurs during renaming operations (https://github.com/Unid\
ata/netcdf-c/issues/597). 4) Attempting to compress or decompress a netCDF4 dataset with a non-standard (i.e., non-DEFLATE) filter when the requisite shared library to encode/decode that compressi\
on filter is not present in either the default location (/usr/local/hdf5/lib/plugin) or in the user-configurable location referred to by the HDF5_PLUGIN_PATH environment variable. One can determin\
e if missing plugin libraries are the culprit by dumping the hidden attributes of the dataset with, e.g., ncks --hdn -m in.nc or ncdump -s -h in.nc. Any variables with (hidden) "_Filter" attribute\
s require the corresponding shared libraries to be located in HDF5_PLUGIN_PATH. 5) Bad vibes.
nco_err_exit(): ERROR NCO will now exit with system call exit(EXIT_FAILURE)
Traceback (most recent call last):
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/ush/merge_fv3_aerosol_tile.py", line 185, in <module>
    main()
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/ush/merge_fv3_aerosol_tile.py", line 179, in main
    merge_tile(out_file_name, ctrl_file_name, core_file_name, rest_file_name, chem_file_name, variable_names)
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/ush/merge_fv3_aerosol_tile.py", line 143, in merge_tile
    subprocess.run(["ncatted", "-a", "checksum,,d,,", base_file_name], check=True)
  File "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/core/miniconda3/4.6.14/envs/pygw/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ncatted', '-a', 'checksum,,d,,', '/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3test3/gfs.20190802/00/atmos/INPUT/gfs_data.tile1.nc.tmp']' returned non-ze\
ro exit status 1.
Traceback (most recent call last):
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/scripts/exgfs_aero_init_aerosol.py", line 158, in <module>
    main()
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/scripts/exgfs_aero_init_aerosol.py", line 64, in main
    merge_tracers(merge_script, atm_files, tracer_files, rest_files, core_files[0], ctrl_files[0], tracer_list_file)
  File "/scratch2/NAGAPE/arl/Barry.Baker/gw_new/scripts/exgfs_aero_init_aerosol.py", line 153, in merge_tracers
    subprocess.run([merge_script, atm_file, tracer_file, core_file, ctrl_file, rest_file, tracer_list_file, temp_file], check=True)
  File "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/core/miniconda3/4.6.14/envs/pygw/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/scratch2/NAGAPE/arl/Barry.Baker/gw_new/ush/merge_fv3_aerosol_tile.py', '/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3test3/gfs.20190802/00/atmos/INPUT/g\
fs_data.tile1.nc', '/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3test3/gfs.20190801/00/atmos/RERUN_RESTART/20190802.000000.fv_tracer.res.tile1.nc', '/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3\
test3/gfs.20190801/00/atmos/RERUN_RESTART/20190802.000000.fv_core.res.nc', '/scratch1/NCEPDEV/stmp2/Barry.Baker/COMROT/fv3test3/gfs.20190802/00/atmos/INPUT/gfs_ctrl.nc', '/scratch1/NCEPDEV/stmp2/B\
arry.Baker/COMROT/fv3test3/gfs.20190801/00/atmos/RERUN_RESTART/20190802.000000.fv_core.res.tile1.nc', '/scratch2/NAGAPE/arl/Barry.Baker/gw_new/parm/chem/gocart_tracer.list', '/scratch1/NCEPDEV/stm\
p2/Barry.Baker/COMROT/fv3test3/gfs.20190802/00/atmos/INPUT/gfs_data.tile1.nc.tmp']' returned non-zero exit status 1.

bbakernoaa commented:
@WalterKolczynski-NOAA adding the `export HDF5_USE_FILE_LOCKING=FALSE` fixed the issue to the `jobs/rocoto/aerosol_init.sh` script seems to have fixed it. I'll check the results in a moment to make sure everything started up correctly.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA I got this compiled and running but it failed on the gfsaerosol_init on the second forecast day. I'm going to keep looking into it but this is the error right now. Specifically, it is failing in the `merge_fv3_aerosol_tile.py` script where the `ncatted` call occurs.

I'd gotten this before, but I thought I had fixed this by adding `base_file.close()`. I'd rather not remove the file lock if we can help it.

bbakernoaa commented:
@WalterKolczynski-NOAA Got you.  I found the spot that is causing the error.  We need to add another `base_file.close()` before the `ncatted` call in the `merge_fv3_aerosol_tile.py` script.

```python
    print("Updating ntracer")

    # Use ncks to rewrite file without ntracer so we can define it anew
    # subprocess.run(["ncks", "-x", "-v", "ntracer", "-O", base_file_name, base_file_name+'2'], check=True)
    subprocess.run(["ncks", "-x", "-v", "ntracer", "-O", base_file_name, base_file_name], check=True)

    base_file = netCDF4.Dataset(base_file_name, "r+")
    base_file.createDimension("ntracer", new_ntracer)

    base_file.close() # <---------------- Needs to be closed again
    # # print(base_file.dimensions["ntracer"])

    # # Remove checksum
    subprocess.run(["ncatted", "-a", "checksum,,d,,", base_file_name], check=True)
```

Doing this we done need to add the HDF environment variable and the files don't get corrupted.

WalterKolczynski-NOAA commented:
Excellent

KateFriedman-NOAA commented on modulefiles/module_base.hera:
```diff
@@ -43,10 +43,11 @@ module load upp/10.0.8
+module use /scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/modulefiles/stack
+module load hpc/1.2.0
+module load hpc-miniconda3/4.6.14
+module load pygw/1.0.0
```
Curious...what is pygw?

WalterKolczynski-NOAA replied:
pygw is the new miniconda virtual environment for GFS

KateFriedman-NOAA commented on parm/config/config.fcst:
```diff
@@ -251,7 +251,7 @@ elif [[ "$CDUMP" == "gfs" ]] ; then # GFS cycle specific parameters
     export restart_interval_gfs=${restart_interval_gfs:-0}
     if [ $restart_interval_gfs -le 0 ]; then
-        export restart_interval=0
+        export restart_interval="$FHMAX_GFS"
```
This is a big change...to go from 0 to 384. What is this changing? This will impact the gfs suite in cycled atmos-only mode.

WalterKolczynski-NOAA replied:
It's actually not. a restart_interval of 0... just writes a restart at the end. The substitution is needed to make the next bit work , because AFAIK having 0 be one of multiple restart_intervals is not the same as 0 by itself.

KateFriedman-NOAA replied:
Ah, right, thanks for explaining! :)

KateFriedman-NOAA commented on parm/config/config.fcst:
```diff
@@ -264,6 +264,12 @@ elif [[ "$CDUMP" == "gfs" ]] ; then # GFS cycle specific parameters
+    if [ $DO_AERO = "YES" ]; then
+        # Make sure a restart file is written at the cadence time
+        if [[ ! "${restart_interval[*]}" =~ "$STEP_GFS"  ]]; then
+            export restart_interval="$STEP_GFS $restart_interval"
```
So for a configuration that runs the gfs suite every cycle out to 384hrs this would be: `restart_interval="6 384"`? Am I understanding this correctly?

WalterKolczynski-NOAA replied:
Yes, but only if aerosols are on.

KateFriedman-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -252,6 +253,7 @@ def get_workflow(dict_configs, cdump='gdas'):
     do_wafs = base.get('WAFSF', 'NO').upper()
     do_vrfy = base.get('DO_VRFY', 'YES').upper()
     do_metp = base.get('DO_METP', 'NO').upper()
+    n_tiles = 6
```
This is the # of FV3 tiles, right? If so, this seems like something we should add into a config file and then source from there (?).

WalterKolczynski-NOAA replied:
Yes and yes, though I was fine hard-coding it since I don't expect it to change unless/until we have to worry about nests.

WalterKolczynski-NOAA replied:
Best location would be config.base, since it pops up a bunch of places. Right now it is hard-coded in all of them.

WalterKolczynski-NOAA replied:
Replacing all the places we have 6 with a centralized variable for number of tiles needs its own PR, if we are going to bother.

KateFriedman-NOAA reviewed (changes requested):
The bulk of the files changed are to convert them from ksh to bash (nice, thanks!). The other changes look ok to me but could you expound on what the new python scripts are doing? Are they for preparing aerosol ICs? Left a few comments.

WalterKolczynski-NOAA commented:
> The bulk of the files changed are to convert them from ksh to bash (nice, thanks!). The other changes look ok to me but could you expound on what the new python scripts are doing? Are they for preparing aerosol ICs? Left a few comments.

The python scripts are actually doing a pretty simple thing: they take the aerosol tracer fields from the restart files of the previous cycle valid at the initial time and add them to the traditional initial conditions for the current cycle. There's also now some modification of the tracers just to maintain mass conservation since the restart file pressure levels can be different than those in the analysis file.

CoryMartin-NOAA commented:
I'll be the one to ask annoying questions unsolicited :-)

Why are we modifying RESTART files?
Should this not just copy RESTART files to continue with aerosols? If we need to update meteorological conditions, an increment should be passed to the model. We can compute increments from the difference between GDAS analysis and the output from the forecast using existing tools.

bbakernoaa commented:
@CoryMartin-NOAA This only effects the aerosol fields.  This is done to ensure mass conservation when the GDAS comes in and modifies the pressure during the cycle run.  The aerosol team has worked very hard to ensure this and if it isn't included aerosols become unstable in the continuous runs

CoryMartin-NOAA commented:
@bbakernoaa I'm confused though, I was under the impression that this is copying tracers from the actual RESTART files and putting them in some new 'fake' RESTART files to the initialize the model. My concern is with these 'fake' RESTART files. Where do these GDAS RESTART files come from? It seems to me we should be passing the model the identical RESTART file it wrote out, and use IAU forcing on the meteorological fields, but perhaps I'm missing something.

WalterKolczynski-NOAA commented:
> I'll be the one to ask annoying questions unsolicited :-)
> 
> Why are we modifying RESTART files? Should this not just copy RESTART files to continue with aerosols? If we need to update meteorological conditions, an increment should be passed to the model. We can compute increments from the difference between GDAS analysis and the output from the forecast using existing tools.

The restart files aren't being modified, the gfs_data files are, and only to add new tracers (existing fields are unchanged). It is also currently only available in forecast-only mode. That said, we did change this in GEFS to use increment files before the offline tracer conservation portion was put in. I'm not sure why @rmontuoro went back to the original concept. We can definitely discuss adopting the GEFS approach in the future, but this has already been vetted by the aerosol team so I'd like to put this in, then change it later if we decide that is appropriate.

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA ok so this is using cold starts with updated tracers? IMO this is not a good approach, and makes it impossible for us to do aerosol DA in the near future. The cold starts vs warm starts issue is probably why Barry has to go through great trouble for mass conservation.

CoryMartin-NOAA commented:
but, @WalterKolczynski-NOAA @bbakernoaa I know this is important for science work ASAP, so I'm okay with it proceeding (for now) with the understanding that we will need to discuss changing this before operational implementation.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA ok so this is using cold starts with updated tracers? IMO this is not a good approach, and makes it impossible for us to do aerosol DA in the near future. The cold starts vs warm starts issue is probably why Barry has to go through great trouble for mass conservation.

Yeah, this is for cold starts. With aerosol DA, we wouldn't need this step at all.

rmontuoro commented:
This feature is required for model evaluation and it will not affect future DA cycling.

CoryMartin-NOAA commented:
> required for model evaluation

I'm worried that any evaluation/tuning that happens from this won't be 100% valid when we transition this to ops because of the differences in meteorological forcing associated with cold starts vs warm starts and the way the model initializes. But perhaps that is overblown/unwarranted.

bbakernoaa commented:
It is my understanding that a feature like this ie already in operation for
GEFS-Aerosol. It is required without DA because we need mass conservation
from one step to the next if we don’t have DA constraining the mass.

On Tue, Mar 29, 2022 at 4:45 PM Cory Martin ***@***.***>
wrote:

> required for model evaluation
>
> I'm worried that any evaluation/tuning that happens from this won't be
> 100% valid when we transition this to ops because of the differences in
> meteorological forcing associated with cold starts vs warm starts and the
> way the model initializes. But perhaps that is overblown/unwarranted.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1082356834>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIUVN2VRJ4LXHVUZ2JZZ4DVCNTWXANCNFSM5RSOU72A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

CoryMartin-NOAA commented:
@bbakernoaa no I'm 95% sure that GEFS-Aerosol uses warm-start RESTART files and doesn't touch the aerosol fields at all and computes meteorology increments as the difference between GDAS and the aerosol member meteorology and passes them to the model.

bbakernoaa commented:
You are right but this is essentially doing the exact same thing. Both are
doing mass conservation between the change in increment. This is essential
for model development and any operational implementation without DA.

On Tue, Mar 29, 2022 at 4:52 PM Cory Martin ***@***.***>
wrote:

> @bbakernoaa <https://github.com/bbakernoaa> no I'm 95% sure that
> GEFS-Aerosol uses warm-start RESTART files and doesn't touch the aerosol
> fields at all and computes meteorology increments as the difference between
> GDAS and the aerosol member meteorology and passes them to the model.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1082362129>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIUVN7K5ED4DHTSEUAOPKTVCNUQNANCNFSM5RSOU72A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

CoryMartin-NOAA commented:
Can we not just use this: https://github.com/NOAA-EMC/GEFS/blob/develop/scripts/exgefs_chem_init.py ?

CoryMartin-NOAA commented:
Perhaps I'm missing something, @bbakernoaa, but the issue of mass conservation is a result of cold starts being 65 layers while warm starts are 64, right? So the GEFS-Aerosol approach removes that concern and simplifies the workflow. I am just still worried about what the model does to go from 65 to 64 in a cold start vs staying at 64 in a warm start.

WalterKolczynski-NOAA commented:
> It is my understanding that a feature like this ie already in operation for GEFS-Aerosol. It is required without DA because we need mass conservation from one step to the next if we don’t have DA constraining the mass.
> […](#)
> On Tue, Mar 29, 2022 at 4:45 PM Cory Martin ***@***.***> wrote: required for model evaluation I'm worried that any evaluation/tuning that happens from this won't be 100% valid when we transition this to ops because of the differences in meteorological forcing associated with cold starts vs warm starts and the way the model initializes. But perhaps that is overblown/unwarranted. — Reply to this email directly, view it on GitHub <[#693 (comment)](https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1082356834)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFIUVN2VRJ4LXHVUZ2JZZ4DVCNTWXANCNFSM5RSOU72A> . You are receiving this because you were mentioned.Message ID: ***@***.***>

History lesson time!

When this was first implemented into GEFS, we tried this method, sans the mass conservation adjustment. Without the mass adjustment... it didn't end well. The codes here originated from that initial attempt.

With time very short and a way to conduct the mass conservation not readily available, after much discussion we transitioned to the increment method, which calculates an increment from the analysis files and warm starts from the restart file ([here's the commit where I added it](https://github.com/NOAA-EMC/GEFS/commit/1ad2c7b54a0aec929ab244532781ed3f843c9667)).

So now this returns to that original method, but now with mass conservation included. I don't know enough to say for certain which way is better (though I suspect it is still the increment). But using an increment + warm start would mean it is being run differently with aerosols vs without, and that might be important for comparisons.

In any event, still going to have this one go through since it has been vetted by the aerosol team, but I would recommend a meeting with all of the relevant science principles to hash this out in the near future. We should be able to steal most of the necessary code from GEFS if the decision is to go the increment route.

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA in fact "But using an increment + warm start would mean it is being run differently with aerosols vs without, and that might be important for comparisons." is *exactly* what I am worried about

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA in fact "But using an increment + warm start would mean it is being run differently with aerosols vs without, and that might be important for comparisons." is _exactly_ what I am worried about

Well, as it is right now, with and without will both be starting from gfs_data.tileN.nc files (albeit ones with extra tracers in the aerosol case). Changing it for aerosols like we do in GEFS would mean now warm starting from the restart files with an increment from the atmanl.nc file. So mechanically, it would be different, even if they are just different paths that lead to a similar place.

rmontuoro commented:
I am afraid this conversation is moving in a different direction and overlooking the purpose of this PR.

This feature was included in coupled-CROW and allowed the aerosol team to perform model evaluation and tuning against ATOM datasets during P7 development. The capability was dropped when coupled-CROW was deprecated, but it is still needed for evaluation and tuning of science updates to UFS-Aerosols.

This is just one of the possible ways we can configure and run a coupled aerosol system. It has nothing to do and it will not have anything to do with DA cycling or with how aerosol cycling was implemented for GEFS-Aerosols.

The aerosol team has been patiently waiting for this coupled-CROW feature to be ported to the current global workflow in order to resume their validation experiments for model improvements planned for P8c.

CoryMartin-NOAA commented:
@rmontuoro understood. And I am fine proceeding as is with the understanding it will be revisited later. Agreed, we need this capability ASAP, as the DA team cannot generate background error statistics without it either. Sorry for derailing the conversation.

bbakernoaa commented on ush/merge_fv3_aerosol_tile.py:
```diff
@@ -0,0 +1,185 @@
+    base_file = netCDF4.Dataset(base_file_name, "r+")
+    base_file.createDimension("ntracer", new_ntracer)
+    # print(base_file.dimensions["ntracer"])
+
```
need to add `base_file.close()` here

WalterKolczynski-NOAA committed:
Close NetCDF file when done in merge aerosol

Needed another close of the NetCDF file when the second modification is done.

Refs: #516

zhanglikate commented:
@CoryMartin-NOAA @bbakernoaa  I can clarify and confirm that GEFS-Aerosols is 100% using warm start, not cold start. The current clod start may be a temporal solution in short term before aerosol DA system. When the aerosols DA is implemented, to be consistent as the whole DA system, I think it should switch to warm start finally.

bbakernoaa commented:
@CoryMartin-NOAA @zhanglikate I understand the difference between the two methods.  Regardless the are meant to achieve the same thing, ie mass conservation.  Let's stay focused on what this pull request is about and that is not DA.  We need to run this in a consistent manor with the meteorology in the forecast-only mode.  This does not say anything about the `cycling` mode (where DA is included) which is not addressed in this pull request.

CoryMartin-NOAA reviewed (approved):
Barry's comment about "a consistent manor with the meteorology in the forecast-only mode" is spot on, and is what I was referring to in my comments earlier. This should definitely be merged for that purpose.

bbakernoaa commented:
@WalterKolczynski-NOAA I did just notice one thing.  The gocart.inst_aod did not get copied to the COMROT directory. I think that this should be added back if aerosols is on.

zhanglikate commented:
@WalterKolczynski-NOAA  @bbakernoaa In the ATM only run, the chemical model AOD output has not been archived in the output folder as the S2S run. Because in the ush/forecast_postdet.sh, in the subroutine "GOCART_postdet()", the variables of "COMOUT", "fhrlst" and "IDATE" have not been defined in the ATM only run, they are from the definitions in other components. 
The following is my testing with modifications, which worked well to get the AOD in the output folder.

GOCART_postdet() {
  echo "SUB ${FUNCNAME[0]}: Linking output data for GOCART"
  export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc"
  [[ ! -d $COMOUT/chem ]] && mkdir -p $COMOUT/chem
  fhrlst=$OUTPUT_FH
  export IDATE=$CDATE
  for fhr in $fhrlst; do
    if [ $fhr = 'anl' ]; then
      continue
    fi
    VDATE=$($NDATE $fhr $IDATE)
    YYYY=$(echo $VDATE | cut -c1-4)
    MM=$(echo $VDATE | cut -c5-6)
    DD=$(echo $VDATE | cut -c7-8)
    HH=$(echo $VDATE | cut -c9-10)
    SS=$((10#$HH*3600))

bbakernoaa commented:
@WalterKolczynski-NOAA Just to make it a little easier to see. I can confirm that this works by modifying the `ush/forecast_postdet.sh` file. 

BEFORE: 

```bash 
GOCART_postdet() {
  echo "SUB ${FUNCNAME[0]}: Linking output data for GOCART"

  [[ ! -d $COMOUT/chem ]] && mkdir -p $COMOUT/chem

  for fhr in $fhrlst; do
    if [ $fhr = 'anl' ]; then
      continue
    fi
    VDATE=$($NDATE $fhr $IDATE)
    YYYY=$(echo $VDATE | cut -c1-4)
    MM=$(echo $VDATE | cut -c5-6)
    DD=$(echo $VDATE | cut -c7-8)
    HH=$(echo $VDATE | cut -c9-10)
    SS=$((10#$HH*3600))

    $NLN $COMOUT/chem/gocart.inst_aod.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_aod.${YYYY}${MM}${DD}_${HH}00z.nc4
  done
}

```

AFTER:

```bash
GOCART_postdet() {
  echo "SUB ${FUNCNAME[0]}: Linking output data for GOCART"

  if [ $RUN_ENVIR = "nco" ]; then
    export COMIN=${COMIN:-$ROTDIR/$RUN.$PDY/$cyc}
    export COMOUT=${COMOUT:-$ROTDIR/$RUN.$PDY/$cyc}
  else
    export COMIN="$ROTDIR/$CDUMP.$PDY/$cyc"
    export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc"
  fi

  [[ ! -d $COMOUT/chem ]] && mkdir -p $COMOUT/chem

  fhrlst=$OUTPUT_FH
  export IDATE=$CDATE
  for fhr in $fhrlst; do
    if [ $fhr = 'anl' ]; then
      continue
    fi
    VDATE=$($NDATE $fhr $IDATE)
    YYYY=$(echo $VDATE | cut -c1-4)
    MM=$(echo $VDATE | cut -c5-6)
    DD=$(echo $VDATE | cut -c7-8)
    HH=$(echo $VDATE | cut -c9-10)
    SS=$((10#$HH*3600))

    $NLN $COMOUT/chem/gocart.inst_aod.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_aod.${YYYY}${MM}${DD}_${HH}00z.nc4

  done
}
```

WalterKolczynski-NOAA committed:
Standardize COMIN/COMOUT

Atmos and waves were using COMIN<component> and COMOUT<component>, but
other components were instead using COMOUT and appending the component
directory on-the-fly. Ocean, ice, and aerosols now have and use their
own COMOUT<component>.

Also fixes a bug where certain variables were not set properly for atm-
aerosol forecast jobs because they had been defined in a section only
run for MOM6.

WalterKolczynski-NOAA commented:
Newest commit should solve the atm-aero problem

bbakernoaa commented:
@WalterKolczynski-NOAA We need to make sure that in the ATM with aerosols run we include the `fscav_aero` variable in the namelist via the `parsing_namelist_FV3.sh` script

Right now it seems to only include it if the variable `$cpl` is included.  This of course isn't there with run with just ATM + AERO

```bash
if [ $cpl = .true. ]; then
  cat >> input.nml << EOF
  fscav_aero = ${fscav_aero:-'*:0.0'}
EOF
fi
```

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA We need to make sure that in the ATM with aerosols run we include the `fscav_aero` variable in the namelist via the `parsing_namelist_FV3.sh` script
> 
> Right now it seems to only include it if the variable `$cpl` is included. This of course isn't there with run with just ATM + AERO
> 
> ```shell
> if [ $cpl = .true. ]; then
>   cat >> input.nml << EOF
>   fscav_aero = ${fscav_aero:-'*:0.0'}
> EOF
> fi```
> ```

This appears to be an error I made when I made the final implementation of aerosols by using the --aerosols flag instead of more apps. cpl is actually supposed to be on if any additional component is on, including GOCART. An upcoming commit reverts it back to using the APP and corrects the problem. In the meantime, add `export DO_COUPLED="YES"` to config.base after line 164.

bbakernoaa commented:
@WalterKolczynski-NOAA Thanks Just wanted to bring this to your attention.  Glad there is already something in the works to fix this

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA Thanks Just wanted to bring this to your attention. Glad there is already something in the works to fix this

Thanks for that. I wasn't actually aware of the issue, I just incidentally fixed it as part of other work.

zhanglikate commented:
@WalterKolczynski-NOAA  It would be great if you can help to update the configurations as:  
aero_conv_scav_factors="'*:0.3','so2:0.0','msa:0.0','dms:0.0','nh3:0.4','nh4:0.6','bc1:0.6','bc2:0.6','dust1:0.6','dust2:0.6', 'dust3:0.6','dust4:0.6','dust5:0.6','seas1:0.5','seas2:0.5','seas3:0.5','seas4:0.5','seas5:0.5'"

bbakernoaa commented:
Kate these get updated from the parm/config/config.aero

On Fri, Apr 1, 2022 at 4:32 PM Kate.Zhang-NOAA ***@***.***>
wrote:

> @WalterKolczynski-NOAA <https://github.com/WalterKolczynski-NOAA> It
> would be great if you can help to update the configurations as:
> aero_conv_scav_factors="'*:0.3','so2:0.0','msa:0.0','dms:0.0','nh3:0.4','nh4:0.6','bc1:0.6','bc2:0.6','dust1:0.6','dust2:0.6',
> 'dust3:0.6','dust4:0.6','dust5:0.6','seas1:0.5','seas2:0.5','seas3:0.5','seas4:0.5','seas5:0.5'"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1086299113>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIUVN7TC7FRJQ24ODVEJ7LVC5MN5ANCNFSM5RSOU72A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WalterKolczynski-NOAA commented:
If these changes are part of P8b, they will be updated when I merge that back to this branch (which might be a week or so off). Otherwise, changes should probably be part of the process for P8c.

zhanglikate commented:
@bbakernoaa Yes. I know. It can also be modified in the job submit path. I just want to put them as the default values that we recommend to use in short term for P8 since they the updating from our recent P7 evaluations.

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA agreed. @zhanglikate  hopefully the configuration is consistent with what is in p8b and if it's not we need to make sure the configuration also gets updated in the ufs-waether-model coupled prototype benchmark tests so that we know we're testing the same configuration there.

zhanglikate commented:
@WalterKolczynski-NOAA @JessicaMeixner-NOAA . Got it. Let's not do that for now. Thanks

zhanglikate commented:
@WalterKolczynski-NOAA @bbakernoaa  I am wondering has this issue been fixed or not? In the most recent version that I checkout from ( git clone -b develop https://github.com/bbakernoaa/global-workflow ) , the issue is still there for ATM-only run with aerosols. 

Thanks.

> > @WalterKolczynski-NOAA We need to make sure that in the ATM with aerosols run we include the `fscav_aero` variable in the namelist via the `parsing_namelist_FV3.sh` script
> > Right now it seems to only include it if the variable `$cpl` is included. This of course isn't there with run with just ATM + AERO
> > ```shell
> > if [ $cpl = .true. ]; then
> >   cat >> input.nml << EOF
> >   fscav_aero = ${fscav_aero:-'*:0.0'}
> > EOF
> > fi```
> > ```
> 
> This appears to be an error I made when I made the final implementation of aerosols by using the --aerosols flag instead of more apps. cpl is actually supposed to be on if any additional component is on, including GOCART. An upcoming commit reverts it back to using the APP and corrects the problem. In the meantime, add `export DO_COUPLED="YES"` to config.base after line 164.

bbakernoaa commented:
@zhanglikate please see https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1086287517

zhanglikate commented:
@bbakernoaa @WalterKolczynski-NOAA , Yes, I manually added based on https://github.com/NOAA-EMC/global-workflow/pull/693#issuecomment-1086287517. Just want to remind this has not been updated in the workflow I checked out.

zhanglikate commented:
@WalterKolczynski-NOAA @bbakernoaa 
The restart_interval is 24 hours in my testing case, however, it did not change the restart file name into the following format to satisfy the dependencies in the workflow. 

      cycle 201607150000 exists
      /scratch2/NCEPDEV/naqfc/Kate.Zhang/fv3gfs/comrot/P8_C96_ATM/gfs.20160715/00/atmos/RERUN_RESTART/20160716.000000.fv_core.res.nc does not exist

Here is my solution, which has been tested to work well:
In ush/forecast_postdet.sh, I added the following in the end:
GOCART_out() {
    # Add time-stamp to restart files at FHMAX
      cd $ROTDIR/${CDUMP}.${PDY}/${cyc}/atmos/RERUN_RESTART
    RDATE=$($NDATE +$FHMAX $CDATE)
    rPDY=$(echo $RDATE | cut -c1-8)
    rcyc=$(echo $RDATE | cut -c9-10)
    for file in $(ls * | grep -v 0000); do
      $NLN $file ${rPDY}.${rcyc}0000.$file
    done
}

In the ush/forecast_postdet.sh, added "[[ $cplchm = .true. ]] && GOCART_out" as following:
before:
  esac
  [[ $cplflx = .true. ]] && MOM6_out
  [[ $cplwav = .true. ]] && WW3_out
  [[ $cplice = .true. ]] && CICE_out
  [[ $esmf_profile = .true. ]] && CPL_out

Current:
  esac
  [[ $cplflx = .true. ]] && MOM6_out
  [[ $cplwav = .true. ]] && WW3_out
  [[ $cplice = .true. ]] && CICE_out
  [[ $cplchm = .true. ]] && GOCART_out
  [[ $esmf_profile = .true. ]] && CPL_out

XianwuXue-NOAA commented on scripts/exgfs_aero_init_aerosol.py:
```diff
@@ -0,0 +1,159 @@
+			print(f'{var} = {f"{var}"}')
+
+	atm_files, ctrl_files = get_atm_files(atm_source_path)
+	tracer_files, rest_files, core_files = get_restart_files(time, incr, max_lookback, fcst_length, rot_dir, cdump)
```
The line shows get_restart_files will return three variables: tracer_files, rest_files, core_files, but the definition of get_restart_files only return one variable (file_list or None), for running None, I got the below errors:
```

WARNING: Unable to find restart files, will use zero fields
Traceback (most recent call last):
  File "/scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/UFS/gw_cont_gocart/scripts/exgfs_aero_init_aerosol.py", line 158, in <module>
    main()
  File "/scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/UFS/gw_cont_gocart/scripts/exgfs_aero_init_aerosol.py", line 61, in main
    tracer_files, rest_files, core_files = get_restart_files(time, incr, max_lookback, fcst_length, rot_dir, cdump)
TypeError: cannot unpack non-iterable NoneType object
```

WalterKolczynski-NOAA replied:
This should be fixed now.

zhanglikate commented:
@XianwuXue-NOAA I am using this version ( https://github.com/bbakernoaa/global-workflow ) that Barry may add some updates based on above discussion and did not see the issue you mentioned. You may download this version and check what is the differences. Thanks.

zhanglikate commented:
@WalterKolczynski-NOAA @bbakernoaa  The GBBEPX emission is not correct in the rc file. Please update it, otherwise, when we choose the GBBPEX emission, there will be not fire emission in the simulation. 
In the file of parm/chem/ExtData.gbbepx, the variables in the GBBPEx emission are not correct

Before:
`SU_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none biomass ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
OC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none biomass ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
BC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none biomass ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
EMI_NH3_BB NA  N Y %y4-%m2-%d2t12:00:00 none none biomass ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc`

Updated:
`SU_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none SO2 ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
OC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none OC ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
BC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none none BC ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc
EMI_NH3_BB NA  N Y %y4-%m2-%d2t12:00:00 none none NH3 ExtData/nexus/GBBEPx/GBBEPx_all01GRID.emissions_v003_%y4%m2%d2.nc`

XianwuXue-NOAA commented:
> @XianwuXue-NOAA I am using this version ( https://github.com/bbakernoaa/global-workflow ) that Barry may add some updates based on above discussion and did not see the issue you mentioned. You may download this version and check what is the differences. Thanks.

@zhanglikate Thanks a lot for your suggestions. I just compared and Barry's develop branch merged the "develop" branch g-w/EMC and Walter's branch. And the "get_restart_files" function seems the same in this branch. I will do more tests.

WalterKolczynski-NOAA committed:
Merge develop into feature/continuous_aerosols

WalterKolczynski-NOAA commented:
All these other changes should be opened up as separate issues or included as part of the P8c changes, as they are outside the scope of this PR, which is just implementing continuous aerosols.

XianwuXue-NOAA commented on scripts/exgfs_aero_init_aerosol.py:
```diff
@@ -0,0 +1,159 @@
+
+	if (tracer_files is not None):
+		merge_tracers(merge_script, atm_files, tracer_files, rest_files, core_files[0], ctrl_files[0], tracer_list_file)
+
```
merge_tracers function will merge files to the same file, this will cause rerunning issue, which the variables have already exist.

XianwuXue-NOAA commented on ush/merge_fv3_aerosol_tile.py:
```diff
@@ -0,0 +1,186 @@
+    print("-" * 8 + "+" + "-" * 22 + "+" + "-" * 22 + "+" + "-" * 24)
+    for variable_name in tracers_to_append:
+        variable = append_file[variable_name]
+        base_file.createVariable(variable_name, variable.datatype, base_file["sphum"].dimensions)
```
When rerun the job, then the variables exist, so the running createVarible will cause failed, this is caused Kate's issue. @zhanglikate I have fixed in my branch and will create a new issue for this improvement.

The fix can be be as follows:

```
if variable_name not in base_file.variables.keys():
> base_file.createVariable(variable_name, variable.datatype, base_file["sphum"].dimensions)
```

WalterKolczynski-NOAA committed:
Fix GBBPX variable names

The variable names for GBBPX were incorrectly defined. They have been
updated to the correct names.

Refs: #516

WalterKolczynski-NOAA committed:
Prevent merge tracers from adding duplicate variable

The tracer merge script was blindly trying to add tracer varaibles to files
without first checking if the variables already exist in the file. This
caused failures if the job was rerun because the restart files had already
had tracers added. Now if the variable already exists in the original file,
a new variable isn't added, but the existing values are still overwritten by
the new computed tracer fields.

Refs: #516

WalterKolczynski-NOAA committed:
Alias final restart files

The final restart files have a different naming convention than mid-
forecast restart files, omitting the valid time. This causes issues
to scripts that look for files at a specifc valid time that may or
may not be the final time. Now symbolic links are created using the
filename pattern that includes the valid time that point to these
final restart files. This will allow things like aerosol init to
proceed properly without needing to run longer forecasts than the
forecast cadence.

Refs: #516

bbakernoaa commented:
@WalterKolczynski-NOAA I was thinking about this more.  GOCART has the capability of outputting much more diagnostics.  We should copy those as well.  

If we add something like this for potential outputs will it fail even if the files aren't there? 

```bash
$NLN $COMOUTaero/gocart.inst_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 
$NLN $COMOUTaero/gocart.inst_ca.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ca.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_ni.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ni.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_su.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_su.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_ss_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ss_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_ca_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ca_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_ni_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ni_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_su_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_su_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_2d.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_2d.${YYYY}${MM}${DD}_${HH}00z.nc4 
$NLN $COMOUTaero/gocart.inst_3d.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_3d.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.tavg_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.tavg_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.tavg_2d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_2d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4
$NLN $COMOUTaero/gocart.tavg_3d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_3d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4
```

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA I was thinking about this more. GOCART has the capability of outputting much more diagnostics. We should copy those as well.
> 
> If we add something like this for potential outputs will it fail even if the files aren't there?
> 
> ```shell
> $NLN $COMOUTaero/gocart.inst_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 
> $NLN $COMOUTaero/gocart.inst_ca.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ca.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_ni.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ni.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_su.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_su.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_ss_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ss_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_ca_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ca_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_ni_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_ni_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_su_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_su_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_2d.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_2d.${YYYY}${MM}${DD}_${HH}00z.nc4 
> $NLN $COMOUTaero/gocart.inst_3d.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_3d.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.tavg_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_du_ss.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.tavg_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.inst_du_bin.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.tavg_2d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_2d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4
> $NLN $COMOUTaero/gocart.tavg_3d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4 $DATA/gocart.tavg_3d_rad.${YYYY}${MM}${DD}_${HH}00z.nc4
> ```

We can add whatever files the aerosol team deem necessary to COM. But that would be in a separate PR, and we should consider carefully whether it is necessary to add the files to COM.

bbakernoaa commented:
Well what if we did it more as a configurable option. 

Something like 

```bash 
for {out_var} in ${go-cart-out-vars}
    $NLN $COMOUTaero/gocart.${out_var}.${YYYY}${MM}${DD}${HH}00z.nc4 $DATA/gocart.${out_var}.${YYYY}${MM}${DD}${HH}00z.nc4
```
where `${go-cart-out-vars}` is a configurable variable list of `'inst_du_ss tavg_du_ss'` with `'inst_aod'` as the default.

I'm not sure why we wouldn't add it here. Why not just take care of it instead of creating more work for ourselves down the road?

WalterKolczynski-NOAA committed:
Add documentation for top init aerosol script

Adds a first-pass at documentation for the script-level script for
the aerosol init job.

Refs: #516

WalterKolczynski-NOAA committed:
Update modules for verif on Hera

Verification (specifically MET and MET+) needed additional modules. The met and
metp modules themselves are hacked versions of the ones from contrib that
change the conda prereq to the miniconda3 we are using in g-w. A permanent
solution will need to be found.

WalterKolczynski-NOAA committed:
Update verif to not load own modules

The verification script was loading its own modules instead of using
the ones global-workflow provides. This was causing issues with the
new miniconda environment added for aerosol init. Now the verification
script no longer loads its own modules.

Refs: #516

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/continuous_aerosols

WalterKolczynski-NOAA commented:
Waiting on:
[hpc-stack PR #324](https://github.com/NOAA-EMC/hpc-stack/pull/324)
[hpc-stack Issue #408](https://github.com/NOAA-EMC/hpc-stack/issues/408)

WalterKolczynski-NOAA commented:
This PR will also moot/fix Issue #630

WalterKolczynski-NOAA committed:
Switch python environment to ufswm, extend to Orion & Dell

The existing ufswm python environment contains identical python libraries to
the python environment we had created, so changed the environment to the
existing environment instead of creating a new one.

Also extended the custom stack loads to Orion and Dell so we can proceed with-
out waiting for hpc-stack to add to the official installation.

Refs: #693

WalterKolczynski-NOAA commented:
We're proceeding with setting up custom partial hpc-stack installations as a work-around until the necessary updates are made to the official hpc-stack installations.

bbakernoaa commented:
@WalterKolczynski-NOAA Thanks!

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/continuous_aerosols

WalterKolczynski-NOAA committed:
Fix aerosol init restart finding function

At some point along the line when the get_restart_files function was expanded
to get three types of files, the structure was broken in a way that would only
lookback one cycle. Additionally, the failure state wasn't updated to conform
to the new method signature. Both of these have now been repaired.

Refs: #516

WalterKolczynski-NOAA commented:
@bbakernoaa @zhanglikate There have been a lot of comments on this issue. Are there any other outstanding issues that I missed that belong in _this_ PR? For any other changes (new settings, new files to COM), please open new issues in global-workflow (unless there already is one).

WalterKolczynski-NOAA commented:
Also, this should work on all three current tier-1 machines now (Hera, Orion, Dell), including verification.

bbakernoaa commented:
@WalterKolczynski-NOAA Thanks Walter.... I'm testing this out now

KateFriedman-NOAA commented on modulefiles/module_base.hera.lua:
```diff
@@ -41,9 +41,17 @@ load(pathJoin("w3nco", "2.4.1"))
+
+-- Temporary until official hpc-stack is updated
+prepend_path("MODULEPATH", "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/modulefiles/stack")
+load(pathJoin("hpc", "1.2.0"))
```
Will the official hpc-stack you're waiting for be `1.2.0` or will it stay `1.1.0`?

WalterKolczynski-NOAA replied:
This is only for the temporary custom portion.

KateFriedman-NOAA commented on modulefiles/module_base.hera.lua:
```diff
@@ -41,9 +41,17 @@ load(pathJoin("w3nco", "2.4.1"))
+-- Temporary until official hpc-stack is updated
+prepend_path("MODULEPATH", "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/modulefiles/stack")
+load(pathJoin("hpc", "1.2.0"))
+load(pathJoin("hpc-intel", "18.0.5.274"))
```
If the official hpc-stack you're waiting for is going to be `1.2.0` then perhaps you can update the line at the top of this to be `1.2.0` and then remove these lines since they are the same as above?

WalterKolczynski-NOAA replied:
This is only for the temporary custom portion.

KateFriedman-NOAA commented on modulefiles/module_base.orion.lua:
```diff
@@ -44,8 +44,15 @@ load("contrib")
 
+-- Temporary until official hpc-stack is updated
+prepend_path("MODULEPATH", "/work2/noaa/global/wkolczyn/save/hpc-stack/modulefiles/stack")
+load(pathJoin("hpc", "1.2.0"))
+load(pathJoin("hpc-intel", "2018.4"))
```
Same comment as ones I left for module_base.hera.lua.

KateFriedman-NOAA commented on modulefiles/module_base.wcoss_dell_p3.lua:
```diff
@@ -54,8 +54,14 @@ load(pathJoin("gempak", "7.3.3"))
 
+-- Temporary until official hpc-stack is updated
+prepend_path("MODULEPATH", "/gpfs/dell2/emc/modeling/noscrub/Walter.Kolczynski/save/hpc-stack/modulefiles/stack")
+load(pathJoin("hpc", "1.2.0"))
```
Similar comment to one I left for module_base.hera.lua.

KateFriedman-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -244,7 +250,11 @@ export FHMAX_GFS=$(eval echo \${FHMAX_GFS_$cyc})
 export FHOUT_HF_GFS=${FHOUT_HF_GFS:-1}
-export STEP_GFS="24"
+if (( gfs_cyc != 0 )); then
+    export STEP_GFS=$(( 24 / gfs_cyc ))
```
Hopefully users don't change `gfs_cyc` in their EXPDIR config.base to anything except 0, 1, 2, or 4. The `setup_expt.py` script already constrains this for generating the EXPDIR config.base but nothing stops the user from manually changing it in their generated config.base (e.g. gfs_cyc=3) and having `STEP_GFS` equal 8 (24/3). Do we want to check gfs_cyc = 0, 1, 2, or 4 on line 253 as well or instead?

WalterKolczynski-NOAA replied:
Outside the scope of this PR. All this is doing is fixing the STEP_GFS calculation. This was just a mistake when brining CROW stuff over. Later, I plan to eliminate gfs_cyc entirely and just set STEP_GFS instead.

KateFriedman-NOAA reviewed (approved):
Unable to comment on science-related changes but workflow updates look good, including the change from ksh to bash in job scripts. Left a few comments but approve of this PR from the workflow perspective.

WalterKolczynski-NOAA committed:
Adjust memory for aerosol init

The memory request for aerosol init was higher than allowed on Dell.
However, a value that fits within the Dell limitations was just a bit
too small for Hera. So the value is decreased quite a bit for all
machines, but is slightly larger on Hera.

Refs: #693
